components/server/src/gitlab/gitlab-context-parser.ts

```diff
@@ -214,29 +214,31 @@ export class GitlabContextParser extends AbstractContextParser implements IConte
         const possibleBranches = await this.gitlabApi.run<GitLab.Branch[]>(user, async g => {
             return g.Branches.all(`${owner}/${repoName}`, { search });
         });
-        if (!GitLab.ApiError.is(possibleBranches)) {
-            for (let candidate of possibleBranches) {
-                const name = candidate.name;
-                const nameSegements = candidate.name.split('/');
-                if (segments.length >= nameSegements.length && segments.slice(0, nameSegements.length).join('/') === name) {
-                    branchOrTagObject = { type: 'branch', name, revision: candidate.commit.id };
-                    break;
-                }
+        if (GitLab.ApiError.is(possibleBranches)) {
```

Comment:
this is a quite interesting point to debug log the error

Reply:
What do you think of wrapping this in a try-catch block instead:

https://github.com/gitpod-io/gitpod/blob/e2439d8a7cbfc4174b1997e28a17c59c1181b178/components/server/src/gitlab/gitlab-context-parser.ts#L173-L205

like this:

```typescript
try {
  ...
} catch (e) {
  log.debug("GitLab context parser: Error handle tree context.", e);
  throw e;
}
```

That would catch these errors and add proper logging for all of them in the `getBranchOrTag` function.

Currently, we have something like this in the logs:

![image](https://user-images.githubusercontent.com/24960040/124931793-d7cc9f00-e002-11eb-859d-f7005a039ff6.png)

The try-catch change would replace this with proper logging, right?

Reply:
I just mean that `GitLab.ApiError.is(possibleBranches)` contain the error message and the like. That would be of interest to see for debugging.

Reply:
Yes. All information available in `possibleBranches` API error are now logged here: https://github.com/gitpod-io/gitpod/blob/7785a6c2ca7b58c2ed85d02a13be39b553dad6af/components/server/src/gitlab/gitlab-context-parser.ts#L208